### ISSUE TITLE
Seal twisting

### ISSUE THREAD

**Issue by nitrogain** [the-hyjal-project/bugtracker#333]
**Describe Your Issue**

Without spell batching, seal twisting (changing seals to get the benefit of both seal procs for a very short time, "https://forum.elysium-project.org/topic/24956-seal-twisting-101/" this explains seal twisting in more details) is not possible without spell batching or a custom workaround in Vanilla WoW.
**Steps To Reproduce**

1. Cast Seal of Command and when your swing animation starts, cast another seal and if Seal of Command procs, you should be able to get the benefit of Seal of Command proc, but also the benefit of the other seal for all extra attacks you do.

Spell: https://classic.wowhead.com/spell=
(from highest rank to lowest order)
Seal of Righteousness: 20293, 20292, 20291, 20290, 20289, 20288, 20287, 21084, 20154 (there are two rank one IDs for some reason)
Seal of Command: 20375, 20915, 20918, 20919, 20920
Seal of the Crusader: 20308, 20307, 20306, 20305, 20162, 21082
Seal of Wisdom: 20357, 20356, 20166
Seal of Light: 20349, 20348, 20347, 20165
Seal of Justice: 20164

**Expected Behaviour**

https://us.forums.blizzard.com/en/wow/t/wow-classic-version-1137-ptr-is-now-available/838945/169

-Developers’ notes: The concept of “seal twisting” was only possible due to spell batching. With a greatly reduced batch size, we’re adding a slight buffer when you swap seals to allow you to proc two in the same swing. This was not an intended interaction in Original WoW, but it’s a fun quirk of WoW Classic gameplay, and we want to allow Paladins to retain the ability to do this. Please let us know how this compares on the PTR to how seal twisting currently works on live realms.

According to Blizzard, it was not an intended function in original WoW, but because it's a fun quirk of the class, they made a custom window for seal twisting (430-560ms). It's basically a bug that has been used as a function on servers with spell batching (and on Classic aswell) to help ret paladin DPS or Nightfall uptime and tanking threat/regen/etc. This would significantly help ret paladins in raids to be more than just viable and considering you have to use two seals, you need to use up more mana with this strategy as a drawback to compensate for the dps increase. It's a hard to do trick with a big drawback, but rewarding that feels balanced and opens up tons of new possibilities especially for theorycrafters. This is a Classic-specific workaround unlike the greater blessing bug, but considering it has existed in the original Vanilla WoW aswell due to spell batching, it feels like an acceptable workaround to include here aswell.

(Information regarding the seal twist window's duration)
![unknown](https://user-images.githubusercontent.com/16620596/107070699-f294b080-67e3-11eb-91b1-10752320d453.png)

**Comment by nitrogain** [comment]
Extra information:
Q - was seal twisting actually present in Vanilla?
A - yes it was from patch 1.11 onwards when SoC received it's 0.5 sec delay - According to Theloras

**Comment by PabloMB95** [comment]
Aditionally, blizzard is keeping seal twisting after reducing the spell batchin windows from 400ms to 10 ms. So it's safe the say that it was in the game, it is in classic and even Blizzard like the idea of keeping this interaction.
https://us.forums.blizzard.com/en/wow/t/wow-classic-version-1137-ptr-is-now-available/838945

**Comment by theloras** [comment]
here are a bunch of clips and screenshots showing how it was implemented on the Classic PTR:

https://cdn.discordapp.com/attachments/530489467140833282/806834175457230908/sealtwist.mp4

https://cdn.discordapp.com/attachments/530489467140833282/806943936866156544/Wacky.mp4

https://clips.twitch.tv/ProtectiveAgreeablePonyDxAbomb

https://cdn.discordapp.com/attachments/530489505057341440/807012779445452840/sor_double_twist.PNG

https://www.youtube.com/watch?v=Cnc-TwCAPqQ&feature=youtu.be

https://cdn.discordapp.com/attachments/530489467140833282/806673572478451742/WowClassicT_AmC8Vz1DfA.png

https://cdn.discordapp.com/attachments/530489505057341440/807274190682980392/unknown.png

https://www.twitch.tv/videos/901955402

https://cdn.discordapp.com/attachments/530489467140833282/806810173842128926/tripple_proc.png

https://cdn.discordapp.com/attachments/530489467140833282/806650121944301618/nomacs_EDZEfCzvdw.png

**Comment by Gamemechanicwow** [comment]
Okay, i am done reading. Let me get this right. 

In Vanilla SoC would linger for ~0.5 sec when switching to a different seal.
In Classic all seals linger for ~0.5 sec when switching to  any seal.

**Comment by Darg727** [comment]
In vanilla, SoC casts 0.5 sec after the triggering white hit. This delay let the animation play without interrupting the regular swing animation. Because this delay was longer than the batching window, it allowed the active seal to activate on the SoC proc which acts like a weapon hit. This is in addition to spell batching allowing both seals to be active for the triggering white hit. This is how you get 2 procs of SoBlood in TBC.

Currently in Classic, SoC procs when you switch to a different seal but fails to cast after the delay without SoC being active on you. This also happens to SoR which is currently being delayed until the next batch window.

Blizzard has recently acknowledged the the desire to allow seal twisting after the removal of spell batching on the PTR. On the PTR, Blizzard added a ~0.5 sec lingering when switching to a different seal. This allows your seals to overlap for ~0.5 seconds. For this time period you actually have 2 seals active.

**Comment by punker85** [comment]
This discussion is full of slight bits of misinformation and you all should know not to trust any "developer notes" left by the team that created Classic WoW as they are out of touch with the design aspects of the vanilla version of the game.  Nothing about spell batching affected this interaction and if you look at the logic of this claim and the particular flow of events that would happen in a paladin's seal-twisting process have no merit.

First, to "twist seals", a paladin would have to prepare to attack an enemy (i.e. before swinging the weapon) with a proc-type seal buff (i.e. Seal of Command or Seal of Justice) that was already active. Then, a **few hundred milliseconds** before the swing animation occurred, the paladin could activate the other seal that he would twist into his upcoming weapon swing/attack. The new seal buff icon would instantly replace the old seal buff icon on the **client** side which meant it was not batched and delay by the server.  Next, the swing animation occurs followed by a short delay (150ms to 250ms) and then the white damage occurs. If the Seal of Command attack proc occurred, then the twisted seal that was activated just before swing would also activate its proc (Seal of Justice was the only other seal with a proc at this time). The **main reason** that this worked is due to a delay between the occurrence of a swing animation in the client (i.e. swing/attack packet is sent to the server) and the damage being dealt in the client (i.e. damage dealt packet received by the client). Private servers do not have this melee attack delay as you will notice when you see the white damage floating upwards from your targets as soon as your attack animation begins.  

On official vanilla server-client architecture, the client first would be sending out a packet to tell the server to activate a new seal aura for the paladin. Then, a few hundred milliseconds later the client would be sending out an attack swing packet that the server must process with some delay before confirming that damage has been done. During this period of delay, the server is processing and calculating how damage will occur with the attack including analysis of the effect of the first seal aura and the new one that was just received/processed in a recent packet that could have arrived slightly before or after the swing attack packet was received by the server. Next, the server sends all of the damage events in a few packets back to the paladin's client. Somewhere in this mix the first seal aura is deactivated server-side but most likely after the attack confirmation packet is sent to the client from the server.  Many of the mechanics involved in this area of the game are not typically discussed in private server development blogs which is why it gets pushed to the wayside.

Here is an old video of the mechanic in action: https://www.youtube.com/watch?v=w5YpDd4z7gw

TL;DR It really has nothing to do with spell-batching but instead a rather obscure mechanic that caused a slight delay between attack animations and damage dealt from the client perspective.

**Comment by Gamemechanicwow** [comment]
It has been implemented.

Whether this effect was caused by batching or not seems to be a different discussion. But there is some evidence that proves it was. Like the SoC procdamage being delayed and dealt on next cycle.

>  the client would be sending out an attack swing packet
Such packet doesn't exist.

**Comment by punker85** [comment]
> It has been implemented.
> 
> Whether this effect was caused by batching or not seems to be a different discussion. But there is some evidence that proves it was. Like the SoC procdamage being delayed and dealt on next cycle.
> 
> > the client would be sending out an attack swing packet
> 
> Such packet doesn't exist.

A spell batching delay for socomm proc does almost nothing for this mechanic in vanilla since socomm was almost exclusively twisted with justice for the free stun that was granted on the socomm proc.  The only thing that makes it relevant is for seal of blood affecting delayed socomm proc when it was twisted in. That still has no bearing on why the white attack gets the effect of both seals.

Also, CMSG_ATTACKSWING has been a packet opcode in mangos distributions since its inception. It's even there in the vmangos repo - https://github.com/vmangos/core/blob/4ba6a87d197242aa5546ddbf438589b23efd9704/src/game/Protocol/Opcodes.cpp#L387

**Comment by Gamemechanicwow** [comment]
> Also, CMSG_ATTACKSWING has been a packet opcode in mangos distributions since its inception. It's even there in the vmangos repo - https://github.com/vmangos/core/blob/4ba6a87d197242aa5546ddbf438589b23efd9704/src/game/Protocol/Opcodes.cpp#L387

That packet does not do what you describe.

**Comment by punker85** [comment]
> > Also, CMSG_ATTACKSWING has been a packet opcode in mangos distributions since its inception. It's even there in the vmangos repo - https://github.com/vmangos/core/blob/4ba6a87d197242aa5546ddbf438589b23efd9704/src/game/Protocol/Opcodes.cpp#L387
> 
> That packet does not do what you describe.

Well, there must be someway to cause this perceived delay between animation and damage with SMSG_ATTACKERSTATEUPDATE that has been previously undiscovered (e.g. through the use of its unused fields). Otherwise, maybe the "obsolete" opcodes have some use after all.
